### ISSUE TITLE
Allow option for regular unix socket

### ISSUE THREAD

**Issue by chiraag-nataraj** [Irqbalance/irqbalance#72]
Could there be a command-line option to allow binding to a regular (path-based) unix socket instead of the abstract socket currently used? Abstract sockets are a nightmare for security when using sandboxing programs like [firejail](https://github.com/netblue30/firejail).

**Comment by nhorman** [comment]
I'm not sure what you mean.  The sun_path in the unix socket creation is specified to be irqbalance<pid>.sock .  Are you asking instead to have a fifo created on the filesystem to facility communication between irqbalance-ui and irqbalance?

**Comment by chiraag-nataraj** [comment]
Yes. That is, a socket with `sun_path[0] != '\0'`.

**Comment by chiraag-nataraj** [comment]
This would facilitate e.g. blacklisting it without having to create a new network namespace (the current solution). Most other programs which create abstract sockets allow the user to switch to using a traditional Unix socket instead (e.g. X11, D-Bus).

**Comment by nhorman** [comment]
You missed what I said.  sun_path is not NULL, its set to irqbalance<pid>.sock.  That should be pretty clear in the code.  Check the init_socket and init_connection functions

**Comment by chiraag-nataraj** [comment]
Something doesn't make sense here. You're right that `sun_path` doesn't start with null. But very clearly (looking at e.g. `sudo ss -xlpn`), the socket being created is an abstract one, and the man page documents as such (at least in Debian).

**Comment by nhorman** [comment]
Looks right to me:
[root@hmswarspite nhorman]# ss -xlpn | grep 1168
u_strLISTEN 0      1                                 @irqbalance1168.sock 22481                                           * 0                                    users:(("irqbalance",pid=1168,fd=5))

the sun_path is listed correctly there by my read

**Comment by chiraag-nataraj** [comment]
The '@' at the beginning means it's an abstract socket.

**Comment by ppwaskie** [comment]
Inside your network namespace, do you bring lo up before trying to test?  If lo is down (which is by default when the namespace is created) then no sockets can be bound to in the namespace that aren't backed by a file on the filesystem.

**Comment by chiraag-nataraj** [comment]
And that's my point. A unix socket that isn't backed by a file on the file system is an abstract socket, which is what I claimed was happening in the OP. And yes, bringing `lo` up does indeed allow me to access the socket. My point is that I'd like the option to use a regular socket rather than an abstract one - I _want_ to be able to blacklist the irqbalance socket without having to create a new network namespace.

**Comment by chiraag-nataraj** [comment]
The nice thing with file-backed sockets is that they're easy to blacklist, even if you can't (for whatever reason) set up a network namespace - you can just blacklist access to a specific folder or file on the filesystem and now the socket is inaccessible. As it currently stands, there is an issue in firejail with creating a network namespace which allows internet access when your connection is wireless, which means I am currently unable to forbid access to the socket unless my program doesn't need internet access at all (which is true for some programs but not all). Sure, that's an issue in firejail (and we're working on figuring it out). But when the only way to hide an abstract socket is to setup a new network namespace, that is a lot more intrusive than simply blacklisting a file or directory.

TL;DR: file-backed sockets mean easier, more granular blacklisting and whitelisting of things.

**Comment by ppwaskie** [comment]
I don't disagree for the most part.  However, having something like this on the filesystem opens the validation matrix up a bit.  If that file-based socket lives on a network-based share (iSCSI device, NFS, etc.), now we need to do something to handle things if it goes away.

I'm glad you did conclude that right now the crux of the issue is the software you're using, firejail.  If that can be fixed, then this isn't an issue.  But longer term, I do see your point.  Let me talk with Neil about how we can approach this and still maintain the flexibility we have today without increasing the validation vectors if possible.

**Comment by nhorman** [comment]
@ppwaskie could we dynamically create the fifo in /tmp or /var/run, or some other location that is guaranteed to both be local to the host, and autoremovable on system reboot (I'm thinking here specifically of tmpfs).  That should solve the test matrix issue I think

**Comment by ppwaskie** [comment]
@nhorman we could.  /var/run assumes you're on a systemd model.  tmpfs can get contentious too.  On one of my older Gentoo systems that started life on openrc, and migrated to sysfs, it doesn't use tmpfs in RAM, rather an actual /tmp on the filesystem.  While they should always be there, the can of worms get much bigger to make sure we cover the bases.

**Comment by nhorman** [comment]
What if we did something like the following pseudo code:
struct fstab *tab;
tab = getfsfile("/var/run");
if (tab .fs_fvstupe == "tmpfs")
     /* Use fifo on /var/run */
else
    /* fall back to anon socket */

We can repeat that pattern ad nauseum for as many file systems and locations as we like.  Its a bit verbose, but it does let us filter out network file systems. 

Another option would be to scrap the  use of the fifo all together and setup a mailbox operation with shared memory.  I'm not keen on that idea, but its an option

**Comment by ppwaskie** [comment]
I actually prefer your approach with the pseudo-code.  Try to put it in tmpfs first, and if that doesn't work, then default to the anonymous non-file-backed socket.  It should hit pretty much any permutation of init managers and filesystem layouts and still give an out so we can continue operating.

I would like to avoid overhauling it too much with a mailbox.  That seems like a powerhammer where we need a tiny screwdriver.

**Comment by ppwaskie** [comment]
@nhorman you want to hack on it or shall I?  I might have some cycles sitting overseas at the moment.

**Comment by nhorman** [comment]
@ppwaskie yes please, if you have time take a stab at it,as I'm trying to hammer out some other bugs this week.   thanks!

**Comment by nhorman** [comment]
@ppwaskie any update here?

**Comment by ppwaskie** [comment]
@nhorman I have something in flight.  Sorry for the delay, but yes, I have patches coming.  Working through some dumb platform issues.

**Comment by nhorman** [comment]
@ppwaskie no worries, just wanted to check in.

**Comment by nhorman** [comment]
resolved by commit 19c25ddc5a13cf0b993cdb0edac0eee80143be34

**Comment by chiraag-nataraj** [comment]
Thank you all for the quick turn-around time! :)
